swarm_dial.go

```diff
@@ -281,162 +280,375 @@ func (s *Swarm) dialPeer(ctx context.Context, p peer.ID) (*Conn, error) {
+				ad := pending[addr]
+				err := s.dialNextAddr(ad.ctx, p, addr, resch)
+				if err != nil {
+					dispatchError(ad, err)
+				}
+			}
+
+			nextDial = nil
+			triggerDial = nil
+
+		case res := <-resch:
+			active--
+
+			if res.Conn != nil {
+				connected = true
+			}
+
+			if done && active == 0 {
+				if res.Conn != nil {
+					// we got an actual connection, but the dial has been cancelled
+					// Should we close it? I think not, we should just add it to the swarm
+					_, err := s.addConn(res.Conn, network.DirOutbound)
+					if err != nil {
+						// well duh, now we have to close it
+						res.Conn.Close()
+					}
+				}
+				return
+			}
+
+			ad := pending[res.Addr]
+
+			if res.Conn != nil {
+				// we got a connection, add it to the swarm
+				conn, err := s.addConn(res.Conn, network.DirOutbound)
+				if err != nil {
+					// oops no, we failed to add it to the swarm
+					res.Conn.Close()
+					dispatchError(ad, err)
+					if active == 0 && len(nextDial) > 0 {
+						triggerDial = triggerNow
+					}
+					continue loop
+				}
+
+				// dispatch to still pending requests
+				for _, reqno := range ad.requests {
+					pr, ok := requests[reqno]
+					if !ok {
+						// it has already dispatched a connection
+						continue
+					}
+
+					pr.req.resch <- dialResponse{conn: conn}
+					delete(requests, reqno)
+				}
+
+				ad.conn = conn
+				ad.requests = nil
 
-	sk := s.peers.PrivKey(s.local)
-	logdial["encrypted"] = sk != nil // log whether this will be an encrypted dial or not.
-	if sk == nil {
-		// fine for sk to be nil, just log.
-		log.Debug("Dial not given PrivateKey, so WILL NOT SECURE conn.")
+				continue loop
+			}
+
+			// it must be an error -- add backoff if applicable and dispatch
+			if res.Err != context.Canceled && !connected {
+				// we only add backoff if there has not been a successful connection
+				// for consistency with the old dialer behavior.
+				s.backf.AddBackoff(p, res.Addr)
+			}
+
+			dispatchError(ad, res.Err)
+			if active == 0 && len(nextDial) > 0 {
+				triggerDial = triggerNow
+			}
+		}
 	}
+}
 
-	//////
+func (s *Swarm) addrsForDial(ctx context.Context, p peer.ID) ([]ma.Multiaddr, error) {
 	peerAddrs := s.peers.Addrs(p)
 	if len(peerAddrs) == 0 {
-		return nil, &DialError{Peer: p, Cause: ErrNoAddresses}
+		return nil, ErrNoAddresses
 	}
+
 	goodAddrs := s.filterKnownUndialables(p, peerAddrs)
-	if forceDirect {
+	if forceDirect, _ := network.GetForceDirectDial(ctx); forceDirect {
 		goodAddrs = addrutil.FilterAddrs(goodAddrs, s.nonProxyAddr)
 	}
+
 	if len(goodAddrs) == 0 {
-		return nil, &DialError{Peer: p, Cause: ErrNoGoodAddresses}
+		return nil, ErrNoGoodAddresses
 	}
 
-	if !forceDirect {
-		/////// Check backoff andnRank addresses
-		var nonBackoff bool
-		for _, a := range goodAddrs {
-			// skip addresses in back-off
-			if !s.backf.Backoff(p, a) {
-				nonBackoff = true
-			}
+	return goodAddrs, nil
+}
+
+func (s *Swarm) mergeDialContexts(a, b context.Context) context.Context {
```

Comment:
I really don't like this but let's punt on making any more changes till a followup.

Reply:
ok, fair enough.